### ISSUE TITLE
add code Python examples using nexusformat

### ISSUE THREAD

**Issue by prjemian** [nexusformat/definitions#807]
At [NIAC2020](https://github.com/nexusformat/NIAC/issues/43), it was suggested that we add Python examples using the [nexusformat](https://github.com/nexpy/nexusformat) package.

**Comment by prjemian** [comment]
Example using the [simplest example](https://manual.nexusformat.org/examples/h5py/writer_1_3.html) from the manual:

```python
#!/usr/bin/env python

from nexusformat import nexus as nxf
import numpy as np

BASE = "."
PATH = f"{BASE}/manual/source/examples/h5py"
DATAFILE = f"{PATH}/input.dat"
OUTPUT_FILE = "/tmp/simple.h5"

def main():
    buffer = np.loadtxt(DATAFILE).T
    tthData = buffer[0]                             # float[]
    countsData = np.asarray(buffer[1],'int32')   # int[]

    # create the NeXus structure and write the data file
    nxf.NXdata(countsData, tthData).save(OUTPUT_FILE)

if __name__ == "__main__":
    main()
```

which produces this tree structure:

```
/tmp/simple.h5 : NeXus data file
  @HDF5_Version = 1.10.6
  @file_name = /tmp/simple.h5
  @file_time = 2020-10-15T13:33:20.279239
  @h5py_version = 2.10.0
  @nexusformat_version = 0.5.3
  entry:NXentry
    @NX_class = NXentry
    data:NXdata
      @NX_class = NXdata
      @axes = axis1
      @signal = signal
      axis1:NX_FLOAT64[31] = [17.92608, 17.92591, 17.92575, '...', 17.92108]
      signal:NX_INT32[31] = [1037, 1318, 1704, '...', 1321]
```

same structure, different names, minor differences in attributes

**Comment by vasole** [comment]
I guess the nexusformat NXdata allows to do exactly the same as the original example.

- The dataset names should be the same
- The unit attributes should also appear
- The indices are optional. They should be made optional in the h5py example or added in the nexusformat example.

**Comment by prjemian** [comment]
Agreed.  Both the published manual and this example are not examples of the *simplest NeXus data file*.  While the code above is the simplest nexusformat code to write a data file (which is worthwhile by itself), it has metadata in the *NXroot* and no *units* attributes.  The published code could easily be changed to use the same field and group names.  The `FIELD_indices` attribute is trivial for 1-D data and can be dropped from the *simplest* examples.

It is obvious how to modify the *h5py* example code.
To add units and control the group and field names, the *nexusformat* example needs a few more steps:

```python
#!/usr/bin/env python

from nexusformat import nexus as nxf
import numpy as np

BASE = "."
PATH = f"{BASE}/manual/source/examples/h5py"
DATAFILE = f"{PATH}/input.dat"
OUTPUT_FILE = "/tmp/simple.h5"

def main():
    buffer = np.loadtxt(DATAFILE).T
    tthData = buffer[0]                             # float[]
    countsData = np.asarray(buffer[1],'int32')   # int[]

    # create the NeXus structure and write the data file
    ds_counts = nxf.NXfield(countsData, name="counts", units="counts")
    ds_ttheta = nxf.NXfield(tthData, name="two_theta", units="degrees")
    nxentry = nxf.NXentry(
        nxf.NXdata(
            ds_counts, ds_ttheta, name="data"),
        name="Scan"
    )
    nxentry.save(OUTPUT_FILE)

if __name__ == "__main__":
    main()

```

resulting in this structure:

```
/tmp/simple.h5 : NeXus data file
  @HDF5_Version = 1.10.6
  @file_name = /tmp/simple.h5
  @file_time = 2020-10-16T10:53:36.305473
  @h5py_version = 2.10.0
  @nexusformat_version = 0.5.3
  Scan:NXentry
    @NX_class = NXentry
    data:NXdata
      @NX_class = NXdata
      @axes = two_theta
      @signal = counts
      counts:NX_INT32[31] = [1037, 1318, 1704, '...', 1321]
        @units = counts
      two_theta:NX_FLOAT64[31] = [17.92608, 17.92591, 17.92575, '...', 17.92108]
        @units = degrees
```

**Comment by rayosborn** [comment]
I presume we will talk about this tomorrow, but I think it would be good to give a meaningful example, without being too complicated. I would suggest something like this in the `nexusformat` version:

```
from nexusformat.nexus import *

Qh = NXfield(np.linspace(0, 5.0, 51), name='Qh')
Qk = NXfield(np.linspace(0, 4.0, 41), name='Qk')
H, K = np.meshgrid(Qh, Qk)
data = NXfield(np.exp(-((H-2.5)**2 + (K-2)**2)), name='data')

root = NXroot()
root['entry'] = NXentry()

root['entry/sample'] = NXsample()
root['entry/sample/temperature'] = NXfield(40.0, units='K')
root['entry/sample/mass'] = NXfield(10.0, units='g')

root['instrument'] = NXinstrument(NXmonochromator())
root['instrument/monochromator/energy'] = NXfield(87.1, units='keV')

root['entry/data'] = NXdata(data, (Qk, Qh))
root['entry/data'].set_default()

root['entry/title'] = 'Example NeXus Data (nexusformat)'

root.save('example.nxs', 'w')
```

In `h5py`, it becomes:
```
import datetime
import h5py as h5
import os

Qh = np.linspace(0, 5.0, 51)
Qk = np.linspace(0, 4.0, 41)
H, K = np.meshgrid(Qh, Qk)
data = np.exp(-((H-2.5)**2 + (K-2)**2))

root = h5.File('example.h5', 'w')

root.create_group('entry')
root['entry'].attrs['NX_class'] = 'NXentry'

root['entry'].create_group('sample')
root['entry/sample'].attrs['NX_class'] = 'NXsample'

root['entry/sample'].create_dataset('temperature', data=40.0)
root['entry/sample/temperature'].attrs['units'] = 'K'
root['entry/sample'].create_dataset('mass', data=10.0)
root['entry/sample/mass'].attrs['units'] = 'g'

root['entry'].create_group('instrument')
root['entry/instrument'].attrs['NX_class'] = 'NXinstrument'
root['entry/instrument'].create_group('monochromator')
root['entry/instrument/monochromator'].attrs['NX_class'] = 'NXmonochromator'
root['entry/instrument/monochromator'].create_dataset('energy', data=87.1)
root['entry/instrument/monochromator/energy'].attrs['units'] = 'keV'

root['entry'].create_group('data')
root['entry/data'].attrs['NX_class'] = 'NXdata'
root['entry/data'].create_dataset('Qh', data=Qh)
root['entry/data'].create_dataset('Qk', data=Qk)
root['entry/data'].create_dataset('data', data=data)
root['entry/data'].attrs['signal'] = 'data'
root['entry/data'].attrs['axes'] = ['Qk', 'Qh']

root['entry'].create_dataset('title', data='Example NeXus Data (h5py)')

root['entry'].attrs['default'] = 'data'
root.attrs['default'] = 'entry'

root.attrs['file_name'] = os.path.abspath('example.h5')
root.attrs['file_time'] = datetime.datetime.now().isoformat()
root.attrs['h5py_version'] = h5.version.version
root.attrs['HDF5_Version'] = h5.version.hdf5_version

root.close()
```
I have tested both and they give identical results.

**Comment by vasole** [comment]
You do not need to make things unnecessary cumbersome with the h5py example. You can avoid all the create calls.

At the very least I would remove the create_dataset calls from your example. The calls to create_group could also be avoided by directly creating the datasets and only at the very end creating the attributes. That way one would see that the main interest of nexusformat is to avoid "forgetting" some attributes on a perfect layout while the main advantage of the h5py approach is that no custom API is to be learned. One just needs to now how to deal with dictionaries.

The main gain of nexusformat is to make sure attributes are not forgotten what can be very convenient for a scientist in interactive sessions. I do not see it as convenient for actual data producers where one is developing code that needs access to more advanced features (soft links, external links, virtual datasets) that will in any case force to learn h5py.

I would modify your h5py example removing the create_dataset calls as below

```
import numpy as np
import datetime
import h5py as h5
import os

Qh = np.linspace(0, 5.0, 51)
Qk = np.linspace(0, 4.0, 41)
H, K = np.meshgrid(Qh, Qk)
data = np.exp(-((H-2.5)**2 + (K-2)**2))

root = h5.File('example.h5', 'w')
root.attrs['NX_class'] = 'NXroot'

root.create_group('entry')
root['entry'].attrs['NX_class'] = 'NXentry'

root['entry'].create_group('sample')
root['entry/sample'].attrs['NX_class'] = 'NXsample'

root['entry/sample/temperature'] = 40.0
root['entry/sample/temperature'].attrs['units'] = 'K'
root['entry/sample/mass'] = 10.0
root['entry/sample/mass'].attrs['units'] = 'g'

root['entry'].create_group('instrument')
root['entry/instrument'].attrs['NX_class'] = 'NXinstrument'
root['entry/instrument'].create_group('monochromator')
root['entry/instrument/monochromator'].attrs['NX_class'] = 'NXmonochromator'
root['entry/instrument/monochromator/energy'] = 87.1
root['entry/instrument/monochromator/energy'].attrs['units'] = 'keV'

root['entry'].create_group('data')
root['entry/data'].attrs['NX_class'] = 'NXdata'
root['entry/data/Qh'] = Qh
root['entry/data/Qk'] = Qk
root['entry/data/data'] = data
root['entry/data'].attrs['signal'] = 'data'
root['entry/data'].attrs['axes'] = ['Qk', 'Qh']

root['entry/title'] = 'Example NeXus Data (h5py)'

root['entry'].attrs['default'] = 'data'
root.attrs['default'] = 'entry'

root.attrs['file_name'] = os.path.abspath('example.h5')
root.attrs['file_time'] = datetime.datetime.now().isoformat()
root.attrs['h5py_version'] = h5.version.version
root.attrs['HDF5_Version'] = h5.version.hdf5_version

root.close()
```

**Comment by rayosborn** [comment]
I must admit that I didn't know you could remove create calls in h5py. I'm happy for the most efficient method to be presented either way. 

My argument for recommending nexusformat over h5py only applies to new, inexperienced users, i.e., the people who are reading the documentation for the first time. You are free to ignore the recommendation - it is no threat to or criticism of the way you work. I believe the purpose of the introductory material is to encourage people that the NeXus format is something they want to support because it is intuitively easy to use. That means not requiring them to spend time learning all the attributes they have to apply to make a file standard-conforming. Most people want to spend as little time as possible reading manuals.

**Comment by rayosborn** [comment]
We should also post examples of reading a NeXus file, e.g., to plot the data. I suspect that would make the difference between the two packages more significant (than a factor 2). Using my example, because I had set the `default` attribute, a plot would be:

```
root = nxload('example.nxs')
root.plot()
```
I would be interested in an h5py example, which presumably would require reading the `default` attributes in the NXroot and NXentry groups, then reading the `axes` and `signal` attributes, checking to see if there is a title, and using all of this to construct the arguments to `plt.imshow` (because the program would have to check the rank of the data and choose that over a simple plot). I would be interested in how you would approach this with h5py.

**Comment by prjemian** [comment]
*Plot* is a good example for a demo of the nexusformat (or is it NeXPy) package but I'm reluctant to create a h5py example.  That goes beyond an example to read a NeXus file.

**Comment by prjemian** [comment]
Zoom chat
[meeting_saved_chat.txt](https://github.com/nexusformat/definitions/files/5402706/meeting_saved_chat.txt)

**Comment by benajamin** [comment]
- [2.1.2. BasicWriter.py](https://manual.nexusformat.org/examples/h5py/index.html#writing-the-hdf5-file-using-h5py)
- [2.1.2. BasicReader.py](https://manual.nexusformat.org/examples/h5py/index.html#reading-the-hdf5-file-using-h5py)
-

**Comment by rayosborn** [comment]
The `root.plot()` command works outside of NeXpy. In a regular Python or IPython session, it would throw up a pyplot window.

**Comment by prjemian** [comment]
So `.plot()` is a method from the *nexusformat* package?

**Comment by rayosborn** [comment]
Yes, there is a default plotter in the nexusformat package, which uses Matplotlib. It is overridden in NeXpy so that the plots are within the GUI. You can see an example in the Jupyter notebook I just set up:

https://colab.research.google.com/drive/1niNh3YEpkx-E_nsoAkCUlMBZ_Xx6W_yg?usp=sharing

**Comment by vasole** [comment]
@rayosborn Nice feature!

**Comment by matumot** [comment]
Hello, 

Thank you very much to provide usable python code examples
through Google Golaboratory.

Though python example codes are not in the nexus manual yet,

I read NeXus example for WONI instrument under 
https://manual.nexusformat.org/applying-nexus.html

then I felt that it may be better to prepare python code examples for:

1. How to read/write with data format defined by 
   customized NXDL XML schema file for WONI ?

2. How to make soft data link in nexus data format ?
   (the code example for hard link is already listed by Ben-san
    [externalExample.py]])

3. How to write/read nexus file 
  if multiple NXentry data are involved?

Is it possible?
I appreciate if we have some comments on theses.

**Comment by benajamin** [comment]
we should also work on https://github.com/nexusformat/definitions/issues/742 at the same time.

**Comment by prjemian** [comment]
### NeXus uses HDF5 **hard** links or external links

Point of reference from the manual (https://manual.nexusformat.org/design.html#links):

> NeXus links are HDF5 hard links with an additional `target` attribute. 

NeXus does not use HDF5 _soft_ links.  Only hard links or external links.

link | description
--- | ---
hard | A hard link is a reference from the address dictionary (such as `/entry/data/data`) to the HDF5 stored data object (such as Object Ref. 7464 as described by _HDFView_).  All datasets and groups in a HDF5 file are stored data objects.
external | An external link from a HDF5 file refers to a HDF5 Dataset or a HDF5 group, which resides in another HDF5 file.
soft link |  Soft links contain just the text path rather than a pointer or the address to the object itself.

ref:  https://pythontic.com/hdf5/h5py/links

**Comment by matumot** [comment]
Thank you to clarify the link.  I tried to fix typo in the sentences for python code examples as below.
Sorry for the confusions due to my misunderstandings.

----->
I read NeXus example for WONI instrument under
https://manual.nexusformat.org/applying-nexus.html

then I felt that it may be better to prepare python code examples for:

1. How to read/write with data format defined by
   customized NXDL XML schema file for WONI ?

2. How to make hard link in nexus data format ?
  (the code example for external link is already listed by Ben-san
  [externalExample.py]])

3. How to write/read nexus file
  if multiple NXentry data are involved?
<---

**Comment by rayosborn** [comment]
As @prjemian says, NeXus doesn't recognize soft links. In `nexusformat`, you can add a hard link either by creating an NXlink object with the target path:
```
>>> root['entry/data/polar_angle'] = NXlink(target='/entry/instrument/detector/polar_angle')
```
or by using the `makelink` function to add the object directly. The target is then automatically deduced from the object path.
```
>>> root['entry/data'].makelink(root['entry/instrument/detector/polar_angle'])
```
The `nxlink` property refers back to the link parent.
```
>>> root['entry/data/polar_angle'].nxlink is root['entry/instrument/detector/polar_angle']
True
```

**Comment by vasole** [comment]
Why do you say "NeXus does not recognize soft links"? That leads to understanding something will not work or that they should be avoided.

The use of SoftLinks is the cleanest way to solve issues like where to find the additional information related to a dataset. Workarounds as suggested in https://github.com/nexusformat/definitions/issues/583 disappear and have no reason even to think about them (no target stuff, no suffixes, nothing).

At the ESRF we are systematically using them. Detector data are systematically under ``NXdetector/data`` and any other reference in the file to those data is made via SoftLinks. Therefore whatever input is provided to analysis code concerning those data, the place where to find the rich information associated is known.

**Comment by vasole** [comment]
```
root['entry/data/polar_angle'] = h5py.SoftLink('/entry/instrument/detector/polar_angle')
```

Is it so difficult to make?

Solutions based on creating an attribute (named target or any other thing) to refer to other data are to be avoided because they are simply dangerous due to the risk of modifying or blocking external files. In that sense, I would recommend @rayosborn to profit from his NXlink abstraction to create a SoftLink.

**Comment by rayosborn** [comment]
I didn't mean to suggest that soft links are banned, just not recognized by the standard. The design of NeXus  doesn't assume that they exist. I think this was originally because HDF4 didn't support soft links - it is certainly impossible in XML. That is the reason why we require the 'target' attribute to be added to linked data. With hard links, there is no way of specifying the parent, which is useful for precisely the reason that @vasole suggests - to allow a detector array in an NXdata group to be associated with its corresponding NXdetector group. Whether you think adding `target` attributes is dangerous or not, it has been a part of the standard since the 90s. When `nexusformat` creates a link, it automatically adds the `target` attribute to the parent to make sure the file is standard-conforming.

If ESRF is not using the `target` attribute, then when `nexusformat` reads the file, it will have no way of associating the linked data. I can check for soft links in `nexusformat`, but it is yet another example of having to work around the many ways people write NeXus files using h5py. I have spent years trying to accommodate the many ways people use h5py in violation of the standard. You can see one of the problems in this [issue](https://github.com/nexpy/nexpy/issues/157). It was a file that had a soft link to an object embedded in an external file that didn't exist. I can now read such files without triggering an exception.

This is another reason why I am in favor of a commonly accepted Python API. It ensures a level of compliance that is otherwise impossible. We cannot expect third-party developers to accommodate this variety when they are reading files from multiple sources. If we all live in our own sandbox, only looking at our own files, this is not a problem, but the moment we step outside the box, we are in a minefield as I know from experience in developing `nexusformat`.

**Comment by vasole** [comment]
Nexusformat and NexPy are perfectly able to perform on ESRF files the actions a generic NeXus supporting program is supposed to be able to perform (reading, plotting,...)

For us it is more critical to make sure acquisition files are not locked due to unnecessary access to them.

Associating a target attribute to the parent can end up associating the attribute to the target. The NeXus adoption of solutions based on datasets (i.e. DATASET_errors) instead of using attributes was motivated by that.

**Comment by rayosborn** [comment]
I think there could be an argument, now that we only use HDF5, for revising the requirement to add a `target` attribute, and recommend the use of soft links instead. It would actually make my life easier, because ensuring that links are synchronized in memory is a pain. However, this is a major structural change to the NeXus standard, so it probably can't happen until the 2022 NIAC meeting.

**Comment by rayosborn** [comment]
@vasole, is there somewhere I can download an ESRF NeXus file containing soft links? I will look into what is involved in treating them as NXlink objects without the `target` attribute.

**Comment by vasole** [comment]
Let me check. Any recent LIMA acquisition file should have an NXdata group whose signal points to an NXdetector data.

**Comment by vasole** [comment]
You can take this file. You will see soft links to the data at two places. Inside an NXcollection and inside an NXdata. The NXcollection could also be an NXdata but in our most generic case it could not. 

http://ftp.esrf.fr/pub/bliss/lima.h5

![image](https://user-images.githubusercontent.com/5636251/96883344-9d28d380-1480-11eb-8992-f4d209a820bb.png)

The use of soft links allows us to reach the NXdetector group whatever path the user takes to select the detector data.

As a comment, that file is part of an acquisition and not all of it. That file is accessed from a top master file containing external links and virtual datasets putting together all the parts of the acquisition (motors, counters, different detectors,...) therefore use of attributes and hard links is really dangerous for us.

**Comment by vasole** [comment]
On that file you will also see the use of the default attribute at NXroot level, at NXentry level, at NXintrument level and at NXdetector level. That way each detector can generate its default plot while allowing to have that plot defined when accessing that detector group from a master file containing multiple detectors and/or a different default plot.

**Comment by rayosborn** [comment]
Thanks for the file. In fact, `nexusformat` already half-implements support for soft links. In your example:
```
>>> lima['entry_0000/measurement/data']
NXlink('/entry_0000/instrument/mpx_cdte_22_eh1/data')
```
It it encounters a soft link, it will create an NXlink object that points to the parent, without adding a `target` attribute, which I think is what @vasole wants. The only problem is that it has no support for writing soft links. If you add a NXlink, it will make it a hard link and add the `target` attribute to both the child and the parent. I could possibly add a `soft=True` keyword to the NXlink creation, but for the time being it is not recognized by the NeXus standard.

We should discuss this in the NIAC. I think there is a strong case to be made for allowing soft links, and even for discontinuing use of the target attribute. For example, it would solve the external link problem - you could create a soft link to an external link without needing to add extra attributes to the external file. It would not be hard to enable this in `nexusformat`.

**Comment by vasole** [comment]
It would provide a much better solution to #583 too. The latter should be kept on hold until there is a decision considering soft links. 

I think the use of the target attribute should be limited to XML where there is no risk. In an HDF5-only world, the target attribute should be deprecated because it is unnecessary and dangerous.

**Comment by benajamin** [comment]
@rayosborn can we please get back on topic here. The aim of this issue is to create new code examples to replace the ones currently in the NeXus manual. Could you please create a separate notebook for each of the code examples we want to replace?

**Comment by prjemian** [comment]
As I stated before, it would be a disappointment for these new examples to
*replace* what is in the documentation now. Rather, they would be an
*addition*. To the point they create the same files without be great for
comparison sake.

**Comment by vasole** [comment]
I agree with @prjemian. There was agreement on adding nexusformat examples not on endorsing nexusformat.

Therefore the use of the word replace is unfortunate.

**Comment by benajamin** [comment]
Sorry, I should clarify that I think we should use the notebooks to _both_ update the h5py code _and_ add the nexusformat code.

**Comment by rayosborn** [comment]
I have written possible `nexusformat` and `h5py` versions of BasicWriter.py and BasicReader.py in the [NeXus Examples notebook](https://colab.research.google.com/drive/1niNh3YEpkx-E_nsoAkCUlMBZ_Xx6W_yg?usp=sharing). The BasicWriter.py is, I think, a perfectly fair comparison. 

I expect some pushback from BasicReader.py, because it shows how much work is necessary to make even a simple h5py program that doesn't already know where everything is already. The `nexusformat` API has a lot of utility functions that make printing file trees, extracting signals and axes, identifying default plots, listing all NXdata groups, and even plotting the data extremely easy. None of that is available in h5py, so I don't know how to make a fair comparison of the two. The current BasicReader.py is extremely unrealistic as it assumes that you know the precise names of every object in the file, allowing you to read them without any further introspection. When a user opens a file, that is not typically going to be the case. 

If someone wants to post another suggestion for BasicReader.py that is more realistic than the current version, I am happy to consider writing a `nexusformat` version.

**Comment by benajamin** [comment]
Perhaps we should include a docstring that lists the assumptions made in order to show minimal code. For example:
- The names of all groups in the hierarchy are known and can be hard-coded
- The file doesn't make use of NXsubentry
- The data is known to be a 2D array

**Comment by rayosborn** [comment]
I have amended the BasicReader.py code to cope with data that is different from rank 2. If it's 3, it reads in a 2D layer (it needs refining for rank > 3). For rank == 3 in a regular Python session, the `nexusformat` API will throw an exception, saying that a 2D slice should be selected, so I added a check for that as well. Of course, in NeXpy, it will plot data of any rank.

**Comment by rayosborn** [comment]
I'm not sure why we would produce a reader, in which every group and field in the file have names that are known in advance, unless it is to make the h5py version look less frightening.

**Comment by prjemian** [comment]
In `BasicReader.py` (the one already in the manual), the names are known in advance, as is typical for many who just want to process their own NeXus data files.  That's what the *Basic* part of the name intends, to represent a basic need of the community.  Very basic use of the standard.  1-D data even.

This new reader (which started from [this code](https://github.com/nexusformat/definitions/issues/807#issuecomment-711480895)) is more than basic, as it includes plotting of a 2-D image and additional fields for energy and temperature.  That is a great *addition*.  It is not a revision of the `BasicReader.py` already in the manual.  So, give names that reflect such difference, such as `ExampleImageWriter.py` and `ExampleImageViewer.py`.

I suggest that we assume the names of the fields and groups are known in advance.  An additional example can demonstrate the consequences (of added code complexity) when these names are not known _a priori_.  But that example is for the benefit of those making a more generic data file reader.  The use case for these *beginner* examples should assume the simpler case where the names are already known.

**Comment by vasole** [comment]
@rayosborn @prjemian 

I think both of your objectives deserve to be published. That would allow to compare:

1 - With the names known in advance, what are the advantages/inconveniences of each approach.
2 - If one wants to visualize a default plot, what is the work involved using nexusformat or h5py.

I do not think it is to make h5py less frightening.  My experience is that rarely the predefined plot is all what a scientist wants to access. Therefore the scientists will in any case need to learn how to explore the file. Both approaches come with a completion helper. Nexusformat in form of attributes and h5py with the h5py.enable_ipython_completer() method allowing to complete dictionaries.

In terms of reading comparison, I expect a huge advantage to nexusformat for point 2 and equality for all the rest.

In terms of writing comparison, I expect nexusformat will be preferred for interactive sessions while data producers will prefer to directly access the full power of h5py without an intermediate layer. If nexusformat would have to support read_direct_chunk, write_direct_chunk, virtual datasets... it would end up re-writing h5py on top of h5py or making so heavy access to h5py that one would miss the advantage of the nexusformat simplicity.

**Comment by rayosborn** [comment]
@vasole, thank you for raising some specific points we can discuss. I have two points to make concerning your examples. 

1. The examples you raise are pretty abstruse. The `read_direct_chunk` function is not part of the h5py high-level API and is not even documented in the h5py manual. In any case, if you want to, go ahead and use h5py directly, but I would argue we shouldn't ask a new user to before they are a lot more experienced.
2. Using the `nexusformat` API does not prevent you from accessing more obscure features. When you open a NeXus file, it creates an h5.File object, e.g., root.nxfile, on which you can perform any h5py operations directly. So you could call root.nxfile[path_to_field].id.read_direct_chunk for those operations that need it and still benefit from the convenience of the rest of the API for more routine operations.

There are very few things you can do in h5py that you can't in `nexusformat`, which has support for setting chunk sizes, compression filters, fletcher32 checksums, etc. Those you can't do, you can do directly on the nxfile object. 

Incidentally, we have used the `nexusformat` API to write nearly 200TB of files using program libraries, not interactively. I rather resent the idea that it is only worthy of being used in interactive sessions. It can do both.

**Comment by vasole** [comment]
@rayosborn My point is that using nexusformat will not prevent from having to learn h5py. Even newcomers have to know from the very first day one that they are are using a subset of what is available.

If we would have studied just NAPI or nexusformat, we would be blocked on problems originated from just using a subset of what is available. Virtual Datasets and External links are standard use at the ESRF. It certainly pays off to have asked our developers to learn h5py and not nexusformat or NAPI. Furthermore, we actively contribute to the development of h5py and all the community benefits from it, including nexusformat.

**Comment by rayosborn** [comment]
@vasole, External links are supported. We use them _all_ the time in our applications. Soft links are already added in my development version and will be released soon, but they are still technically a violation of the standard, so you are apparently arguing that people need to learn h5py in order to violate the standard. I have just been criticized for "defining the standard by implementation." Apparently, it is possible to do that when writing with h5py as well.

Virtual datasets are a fairly new part of HDF5. I believe that they would be read transparently by `nexusformat`, but currently not written. I can probably add it, but what I have said over and over again is that it is fine to use h5py if you need it, but I strongly disagree that we should require every NeXus user to learn the intricacies of h5py minutiae. By all means, train your own software developers to use h5py - just don't require new users to learn it until they are more experienced.

What I don't understand is why you recommend that scientists use `nexusformat` at ESRF, for which I  am very grateful, but then object when I suggest that it would be a good recommendation for other users as well. The reasons for recommending it are exactly the same - it is simpler and you will make fewer mistakes.

**Comment by rayosborn** [comment]
Incidentally, I came across a Sphinx extension at https://sphinxcontrib-osexample.readthedocs.io/. It allows you to add code tabs, so if we agree on code examples, we can have a `nexusformat` tab and a `h5py` tab, which the users are free to click on. Perhaps @prjemian can say whether he thinks the extension could work for the NeXus manual. Having one API example follow the other would quickly make the documentation unwieldy, and mixing them up would make it incoherent. This might be the best solution, if it works, because the user could choose what to look at.

**Comment by prjemian** [comment]
worth a try

**Comment by vasole** [comment]
>What I don't understand is why you recommend that scientists use nexusformat at ESRF, for which I am very grateful, but >then object when I suggest that it would be a good recommendation for other users as well. The reasons for recommending >it are exactly the same - it is simpler and you will make fewer mistakes.

I feel as making an audit of nexusformat and it is not my goal.

Another example. Nexusformat implements its own file locking mechanism in parallel of that of HDF5 and it seems to warrant it will work if all the processes accessing the file use nexusformat. It seems to consider that a timeout (of 10 seconds?) for accessing the file is acceptable for the acquisition. Going closer to HDF5 one can privilege writer access while not preventing readers from accessing the file by using native HDF5 locking mechanisms (and I am not talking about SWMR). The current implementation of nexusformat would block access to our acquisition files. Can it be fixed? Yes, of course, reconsidering nexusformat lock mechanism.

My point of view (and as such it is subjective) is that it is not worth for acquisition to learn yet another API that in many cases will just provide the NX_class attribute.

**Comment by rayosborn** [comment]
File locking is optional and turned off by default.

**Comment by vasole** [comment]
How do you prevent locking then? For pure readers or pure writers it is easy. If you want to do both things in a multi threaded environment I see simpler to work with h5py, but it could well be my lack of knowledge of nexusformat.

**Comment by rayosborn** [comment]
If you want file locking, you need to call `nxsetlock(t)` where `t` is the number of seconds after which an access attempt times out. When `t=0`, which is the default value, there is no file locking. Use `nxgetlock()` to return the currently set value. In NeXpy, you can add a call to `nxsetlock` to ~/.nexpy/config.py if you want to turn it on by default.

**Comment by vasole** [comment]
I meant locking in the HDF5 library way of thinking. I think your locking mechanism will effectively lock a file if using 1.10.x preventing access by another program even if open in read only mode.

Perhaps I am not making myself clear.

Unless HDF5 locking is disabled, the first file opening will gain the file lock (either for reading or for writing) preventing any other access to the file during the time it is opened. Therefore I would expect nexusformat to disable HDF5 file locking when opening a file in read only mode and to make sure that HDF5 file locking is enabled when opening a file in a non read only manner (write or append).

**Comment by rayosborn** [comment]
Ah - I see - I was wondering how you knew that we even had file locks in `nexusformat`. As to HDF5 locks, all `nexusformat` IO operations are done within `with` statements, which of course open and then close the files straight away, e.g., `with root.nxfile:`. Multiple operations can be handled in a single `with` statement to avoid unnecessary file opens - we have a reference counter that handles nested `with` statements. I can't find any documentation about disabling HDF5 locks altogether in the h5py manual. Is that a low-level operation? If that is needed, then you can call the required function from the `root.nxfile` object, which is a sub-class of `h5py.File`.

**Comment by vasole** [comment]
At this point  the only available mechanism is to set the HDF5_USE_FILE_LOCKING variable **prior** to open the file. Not being an atomic operation, one needs to protect the whole operation within a thread lock.

**Comment by rayosborn** [comment]
I am busy with my day job at the moment, but I hope to get back to proposing example files in a few days. I thought it might be helpful to post a quick warning about h5py v3.0.0, which was released today. If you write text as a variable-length string, it used to be automatically returned as unicode, but apparently for consistency, they are now returning it as a byte string. If you have always assumed the former, you will now have to do a version check and decode the returned value or use the new `asstr` function. There are more details [here](https://docs.h5py.org/en/latest/strings.html) with a work-around using version-checking [here](https://gist.github.com/takluyver/32512eb5367d11d013029af5fbf1d3e1).

**Comment by prjemian** [comment]
The updates to the documentation are not ready.  This issue is not necessary for the imminent release 2020.10 of the NXDL standard.  (Once this work is merged, the documentation will be updated right away.)  Moving to the next milestone with expectation this will become ready much sooner.
